lib/streamlit/runtime.py

```diff
@@ -0,0 +1,591 @@
+SCRIPT_RUN_CHECK_TIMEOUT: Final = 60
+
+LOGGER: Final = get_logger(__name__)
+
+
+class SessionClientDisconnectedError(Exception):
+    """Raised by operations on a disconnected SessionClient."""
+
+
+class SessionClient(Protocol):
+    """Interface for sending data to a session's client."""
+
+    def write_forward_msg(self, msg: ForwardMsg) -> None:
+        """Deliver a ForwardMsg to the client.
+
+        If the SessionClient has been disconnected, it should raise a
+        SessionClientDisconnectedError.
+        """
+
+
+class RuntimeConfig(NamedTuple):
+    """Config options for StreamlitRuntime."""
+
+    # The filesystem path of the Streamlit script to run.
+    script_path: str
+
+    # The (optional) command line that Streamlit was started with
+    # (e.g. "streamlit run app.py")
+    command_line: Optional[str]
+
+
+class SessionInfo:
+    """Type stored in our _session_info_by_id dict.
+
+    For each AppSession, the Runtime tracks that session's
+    script_run_count. This is used to track the age of messages in
+    the ForwardMsgCache.
+    """
+
+    def __init__(self, client: SessionClient, session: AppSession):
+        """Initialize a SessionInfo instance.
+
+        Parameters
+        ----------
+        session : AppSession
+            The AppSession object.
+        client : SessionClient
+            The concrete SessionClient for this session.
+        """
+        self.session = session
+        self.client = client
+        self.script_run_count = 0
+
+
+class RuntimeState(Enum):
+    INITIAL = "INITIAL"
+    WAITING_FOR_FIRST_SESSION = "WAITING_FOR_FIRST_SESSION"
+    ONE_OR_MORE_SESSIONS_CONNECTED = "ONE_OR_MORE_SESSIONS_CONNECTED"
+    NO_SESSIONS_CONNECTED = "NO_SESSIONS_CONNECTED"
+    STOPPING = "STOPPING"
+    STOPPED = "STOPPED"
+
+
+class Runtime:
+    def __init__(self, config: RuntimeConfig):
+        """Create a StreamlitRuntime. It won't be started yet.
+
+        StreamlitRuntime is *not* thread-safe. Its public methods are
+        generally safe to call only on the same thread that its event loop runs
+        on.
+
+        Parameters
+        ----------
+        config
+            Config options.
+        """
+        # Will be set when we start.
+        self._eventloop: Optional[asyncio.AbstractEventLoop] = None
+
+        self._script_path = config.script_path
+        self._command_line = config.command_line or ""
+
+        # Mapping of AppSession.id -> SessionInfo.
+        self._session_info_by_id: Dict[str, SessionInfo] = {}
+
+        self._state = RuntimeState.INITIAL
+
+        # Set after Runtime.stop() is called. Never cleared.
+        self._must_stop = asyncio.Event()
+
+        # Set when a client connects; cleared when we have no connected clients.
+        self._has_connection = asyncio.Event()
+
+        # Set after a ForwardMsg is enqueued; cleared when we flush ForwardMsgs.
+        self._need_send_data = asyncio.Event()
```

Comment:
These were `tornado`-based synchronization objects in server.py. They mostly work the same here, with one exception (below).